File: Backend-Folder/backend/app/routes/inspections.py
from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.services.inspection_service import inspection_service

router = APIRouter(prefix="/inspections", tags=["Inspections"])

@router.post("/dispatch")
async def dispatch_inspection(request: Request, db: Session = Depends(get_db)):
    try:
        body = await request.json()
        inst_id = body.get("institutionId")
        if not inst_id:
            raise HTTPException(status_code=400, detail="institutionId is required")
        
        inspector_name = body.get("inspectorName") or "Field Inspector Priya Nair (INSP-492)"
        urgency = body.get("urgency") or body.get("priority") or "Immediate"

        result = inspection_service.dispatch_surprise_inspection(
            db=db, institution_id=inst_id, inspector_name=inspector_name, urgency=urgency
        )
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/submit")
async def submit_inspection(request: Request, db: Session = Depends(get_db)):
    try:
        body = await request.json()
        inst_id = body.get("institutionId")
        observed_staff = body.get("observedStaff") if body.get("observedStaff") is not None else body.get("verifiedStaff")
        
        if not inst_id or observed_staff is None:
            raise HTTPException(status_code=400, detail="institutionId and observedStaff are required")

        result = inspection_service.submit_inspection(db=db, payload=body)
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

File: Backend-Folder/server/routes/inspections.ts
import { Router } from 'express';
import { db } from '../db';

const router = Router();

// POST /api/inspections/dispatch - Dispatch surprise audit
router.post('/dispatch', (req, res) => {
  try {
    const { institutionId, inspectorName, urgency } = req.body;
    if (!institutionId) {
      return res.status(400).json({ success: false, error: 'institutionId is required' });
    }

    const result = db.dispatchSurpriseInspection(
      institutionId,
      inspectorName || 'Field Inspector Priya Nair (INSP-492)',
      urgency || 'Immediate'
    );

    res.json({
      success: true,
      message: `Surprise audit dispatched for ${result.institution.name}`,
      data: result,
    });
  } catch (err: any) {
    res.status(400).json({ success: false, error: err.message });
  }
});

// POST /api/inspections/submit - Submit completed field inspection
router.post('/submit', (req, res) => {
  try {
    const {
      institutionId,
      inspectorId,
      inspectorName,
      observedStaff,
      observedBeneficiaries,
      facilityObservedStatus,
      notes,
      evidencePhotoUrls,
      inspectorLat,
      inspectorLng,
    } = req.body;

    if (!institutionId || observedStaff === undefined) {
      return res.status(400).json({
        success: false,
        error: 'institutionId and observedStaff are required',
      });
    }

    const result = db.submitInspection({
      institutionId,
      inspectorId: inspectorId || 'INSP-492',
      inspectorName: inspectorName || 'Inspector Priya Nair',
      observedStaff: Number(observedStaff),
      observedBeneficiaries: observedBeneficiaries !== undefined ? Number(observedBeneficiaries) : undefined,
      facilityObservedStatus,
      notes: notes || 'Inspection completed via mobile application.',
      evidencePhotoUrls: Array.isArray(evidencePhotoUrls) ? evidencePhotoUrls : [],
      inspectorLat: Number(inspectorLat) || 19.033,
      inspectorLng: Number(inspectorLng) || 73.0297,
    });

    res.json({
      success: true,
      message: 'Inspection submitted successfully and cryptographically recorded.',
      data: result,
    });
  } catch (err: any) {
    res.status(400).json({ success: false, error: err.message });
  }
});

export default router;
